server/rooms.ts

```diff
@@ -1363,17 +1366,51 @@ export class GlobalRoomState {
 			// kill server in 10 seconds if it's still set to
 			setTimeout(() => {
 				if (Config.autolockdown && Rooms.global.lockdown === true) {
-					// finally kill the server
-					process.exit();
+					// finally save state / kill the server
+					void this.kill();
 				} else {
 					this.notifyRooms(
 						notifyPlaces,
-						`|html|<div class="broadcsat-red"><b>Automatic server lockdown kill canceled.</b><br /><br />In the last final seconds, the automatic lockdown was manually disabled.</div>`
+						`|html|<div class="broadcast-red"><b>Automatic server lockdown kill canceled.</b><br /><br />In the last final seconds, the automatic lockdown was manually disabled.</div>`
 					);
 				}
 			}, 10 * 1000);
 		}
 	}
+	async kill() {
+		if (Config.persistbattles) await this.writeBattleState();
+		return process.exit();
+	}
+	async writeBattleState() {
+		const buffer: AnyObject = {};
+		const promises = [];
+		for (const [id, room] of Rooms.rooms) {
+			if (!room.battle) continue;
+			const formatid = room.battle.format;
+			if (!buffer[formatid]) buffer[formatid] = {};
+			const promise = room.battle.getLog().then((log) => {
+				buffer[formatid][id] = {
+					inputLog: log.map(item => item.replace(/\r/g, '')).join('\n'),
+					title: room.title,
+					num: room.roomid.split('-')[2],
```

Comment:
Reminder that storing the `roomid` would simplify a lot of this code and doesn't seem to have disadvantages.